src/socket-proxy/socket-proxyd.c

```diff
@@ -297,11 +300,87 @@ static int connection_enable_event_sources(Connection *c) {
         return 0;
 }
 
+static int send_proxy_protocol_v1(Connection *c) {
+        _cleanup_free_ char *s = NULL;
+        int r, s_len;
+        union sockaddr_union local_addr, remote_addr;
+        socklen_t addr_len;
+
+        r = sd_is_socket(c->server_fd, AF_UNSPEC, SOCK_STREAM, /* listening= */ 0);
+        if (r < 0) {
+                log_warning_errno(errno, "Failed to issue SO_TYPE: %m");
+                goto unknown;
+        }
+        if (r == 0) {
+                log_warning("Only TCP is supported by the PROXY protocol.");
+                goto unknown;
+        }
+
+        addr_len = sizeof(local_addr);
+        if (getsockname(c->server_fd, (struct sockaddr*) &local_addr, &addr_len) < 0) {
+                log_warning_errno(errno, "Failed to get local address (getsockname): %m");
+                goto unknown;
+        }
+
+        addr_len = sizeof(remote_addr);
+        if (getpeername(c->server_fd, (struct sockaddr*) &remote_addr, &addr_len) < 0) {
+                log_warning_errno(errno, "Failed to get remote address (getpeername): %m");
```

Comment:
we usually want that log messages about errors we handle gracefully give a hint how we handle these failures. i.e. in this case maybe suffix with `…, reporting fallback proxy data: %m" or so.